src/LegalizeVectors.cpp

```diff
@@ -0,0 +1,591 @@
+
+    vector<const Let *> lets;
+    Expr visit(const Let *op) override {
+        for (const Let *existing : lets) {
+            internal_assert(existing->name != op->name)
+                << "Let " << op->name << " = ...  cannot be lifted to LetStmt because the name is not unique.";
+        }
+        lets.push_back(op);
+        return mutate(op->body);
+    }
+
+public:
+    Stmt mutate(const Stmt &s) override {
+        ScopedValue<decltype(lets)> scoped_lets(lets, {});
+        Stmt mutated = IRMutator::mutate(s);
+        for (const Let *let : reverse_view(lets)) {
+            mutated = LetStmt::make(let->name, let->value, mutated);
+        }
+        return mutated;
+    }
+
+    Expr mutate(const Expr &e) override {
+        return IRMutator::mutate(e);
+    }
+};
+
+class ExtractLanes : public IRMutator {
+    using IRMutator::visit;
+
+    int lane_start;
+    int lane_count;
+
+    Expr extract_lanes_from_make_struct(const Call *op) {
+        internal_assert(op);
+        internal_assert(op->is_intrinsic(Call::make_struct));
+        vector<Expr> args(op->args.size());
+        for (int i = 0; i < int(op->args.size()); ++i) {
+            args[i] = mutate(op->args[i]);
+        }
+        return Call::make(op->type, Call::make_struct, args, Call::Intrinsic);
+    }
+
+    Expr extract_lanes_trace(const Call *op) {
+        auto event = as_const_int(op->args[6]);
+        internal_assert(event);
+        if (*event == halide_trace_load || *event == halide_trace_store) {
+            debug(3) << "Extracting Trace Lanes: " << Expr(op) << "\n";
+            const Expr &func = op->args[0];
+            Expr values = extract_lanes_from_make_struct(op->args[1].as<Call>());
+            Expr coords = extract_lanes_from_make_struct(op->args[2].as<Call>());
+            const Expr &type_code = op->args[3];
+            const Expr &type_bits = op->args[4];
+            int type_lanes = *as_const_int(op->args[5]);
+            const Expr &event = op->args[6];
+            const Expr &parent_id = op->args[7];
+            const Expr &idx = op->args[8];
+            int size = *as_const_int(op->args[9]);
+            const Expr &tag = op->args[10];
+
+            int num_vecs = op->args[2].as<Call>()->args.size();
+            internal_assert(size == type_lanes * num_vecs) << Expr(op);
+            vector<Expr> args = {
+                func,
+                values, coords,
+                type_code, type_bits, Expr(lane_count),
+                event, parent_id, idx, Expr(lane_count * num_vecs),
+                tag};
+            Expr result = Call::make(Int(32), Call::trace, args, Call::Extern);
+            debug(4) << "  => " << result << "\n";
+            return result;
+        }
+
+        internal_error << "Unhandled trace call in LegalizeVectors' ExtractLanes: " << *event << legalization_error_guide << "\n"
+                       << "Please report this error on GitHub." << legalization_error_guide;
+        return Expr(0);
```

Comment:
I don't think you need "Please report this error on GitHub" because that applies to all internal errors.